code/modules/GOAP/cyka_blyat/actions.dm

```diff
@@ -0,0 +1,339 @@
+/datum/goap_action/russian
+	var/saytext = "Cyka" // Russians can speak so why not!
+
+/datum/goap_action/russian/Perform(mob/living/simple_animal/hostile/russian/A)
+	A.say(saytext)
+	return ..()
+
+/datum/goap_action/russian/attack
+	name = "Attack"
+	cost = 5 // reload + firing = 3, new mag + reload + firing = 4, so this comes if you can't fire/reload+fire/grab mag+reload+fire
+	saytext = "Moving in to melee!"
+
+/datum/goap_action/russian/attack/New()
+	..()
+	preconditions = list()
+	preconditions["enemyDead"] = FALSE
+	effects = list()
+	effects["enemyDead"] = TRUE
+
+/datum/goap_action/russian/attack/AdvancedPreconditions(atom/agent, list/worldstate)
+	var/list/viewl = view(10, agent)
```

Comment:
Especially for instances of looking for mobs nearby to attack, it would be much more CPU efficient to just use loop through one of the many GLOB lists, `carbon_list` would do here, see /code/_globalvars/lists/mobs.dm for more lists

Reply:
So, I tested with this solution a major issue is that AI was able to see through walls which causes pathfinding nightmares, also when profiling it the performance improvement wasn't very massive